certbot/hooks.py

```diff
@@ -57,30 +57,70 @@ def validate_hook(shell_cmd, hook_name):
 
 
 def pre_hook(config):
-    "Run pre-hook if it's defined and hasn't been run."
+    """Run pre-hooks if they exist and haven't already been run.
+
+    When Certbot is running with the renew subcommand, this function
+    runs any hooks found in the config.renewal_pre_hooks_dir (if they
+    have not already been run) followed by any pre-hook in the config.
+    If hooks in config.renewal_pre_hooks_dir are run and the pre-hook in
+    the config is a path to one of these scripts, it is not run twice.
+
+    :param configuration.NamespaceConfig config: Certbot settings
+
+    """
+    if config.verb == "renew":
+        for hook in list_hooks(config.renewal_pre_hooks_dir):
+            _run_pre_hook_if_necessary(hook)
+
     cmd = config.pre_hook
-    if cmd and cmd not in pre_hook.already:
-        logger.info("Running pre-hook command: %s", cmd)
-        _run_hook(cmd)
-        pre_hook.already.add(cmd)
-    elif cmd:
-        logger.info("Pre-hook command already run, skipping: %s", cmd)
+    if cmd:
+        _run_pre_hook_if_necessary(cmd)
 
 pre_hook.already = set()  # type: ignore
 
 
+def _run_pre_hook_if_necessary(command):
+    """Run the specified pre-hook if we haven't already.
+
+    If we've already run this exact command before, a message is logged
+    saying the pre-hook was skipped.
+
+    :param str command: pre-hook to be run
+
+    """
+    if command in pre_hook.already:
+        logger.info("Pre-hook command already run, skipping: %s", command)
+    else:
+        logger.info("Running pre-hook command: %s", command)
+        _run_hook(command)
+        pre_hook.already.add(command)
+
+
 def post_hook(config):
-    """Run post hook if defined.
+    """Run post-hooks if defined.
+
+    This function also registers any executables found in
+    config.renewal_post_hooks_dir to be run when Certbot is used with
+    the renew subcommand.
+
+    If the verb is renew, we delay executing any post-hooks until
+    :func:`run_saved_post_hooks` is called. In this case, this function
+    registers all hooks found in config.renewal_post_hooks_dir to be
+    called followed by any post-hook in the config. If the post-hook in
+    the config is a path to an executable in the post-hook directory, it
+    is not scheduled to be run twice.
+
+    :param configuration.NamespaceConfig config: Certbot settings
 
-    If the verb is renew, we might have more certs to renew, so we wait until
-    run_saved_post_hooks() is called.
     """
 
     cmd = config.post_hook
     # In the "renew" case, we save these up to run at the end
     if config.verb == "renew":
-        if cmd and cmd not in post_hook.eventually:
-            post_hook.eventually.append(cmd)
+        for hook in list_hooks(config.renewal_post_hooks_dir):
+            _run_eventually(hook)
+        if cmd:
```

Comment:
I see what you're getting at, but I'm not sure if I agree.

I would feel differently if pre and post hooks before this PR already ran in a last-in, first-out order. That's not what we do though. When `certbot renew` is run, pre-hooks for each lineage are run in the order they are encountered and the same is done for post-hooks. Each post-hook is saved and they are run before Certbot exits in the order they were found.

A similar argument could (and maybe should) have been made before we shipped the existing hooks system. Pre and post hooks from different configs could also interact with each other in strange ways.

We have the system we currently have though and while I know you're not proposing changing this part, I think it changing the order like this adds some (minor) additional complexity and inconsistency to our hooks. It's another detail that has to be explained and understood. It's an exception to the simple explanation "directory hooks run before config hooks".

I don't have a super strong opinion here, but I do have a preference for what I have in this PR.

@ohemorange, do you have an opinion here? If I'm in the minority here, I'll happily make this change.

Reply:
I see your point, and I don't feel strongly about this either if and when we don't have an actual order for hooks in general. It's just how I see the hooks "layered", and this approach would feel more intuitive for me. But I'm open to other views on the subject as well.

Reply:
I think both options have valuable arguments in the theoretical. What are the ways we expect people will use this feature, and how will either option affect it in practice?

Reply:
I think in the vast majority of cases it won't matter and there will be no interactions between hooks. I'd personally go so far and say it may never matter. For what it's worth, I don't think anyone has complained about similar interactions between hooks from different configs.

In the cases it does matter though, I think people can work around the problem by writing hooks that are:

1. Written defensively
2. Limit unintended/unexpected side effects on the system.

Reply:
I've thought about this, and my only opinion is pick one and document it well in the readthedocs.

Reply:
OK. I still have my initial preference but I don't think it matters that much and I promised we'd land this for 0.19.0.

@joohoi, you can make the final call. Please read and consider my arguments against making this change, but if you also still have your initial preference, let me know and I'll change this first thing tomorrow.

Reply:
Yeah, I agree with you, this is rather minor thing and should be pretty rare corner case that can be covered in the documentation. It boils down to being the best we can, and right now it's being in time. We can leave the functionality as it is in this PR.